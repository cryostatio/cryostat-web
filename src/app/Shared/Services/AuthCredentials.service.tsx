/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Locations } from '@app/Settings/Config/CredentialsStorage';
import { getFromLocalStorage } from '@app/utils/LocalStorage';
import { Observable, of } from 'rxjs';
import type { ApiService } from './Api.service';
import { Credential } from './service.types';

export class AuthCredentials {
  // TODO replace with Redux?
  private readonly store = new Map<string, Credential>();

  constructor(private readonly api: () => ApiService) {}

  setCredential(targetId: string, username: string, password: string): Observable<boolean> {
    const location = getFromLocalStorage('CREDENTIAL_LOCATION', Locations.BACKEND.key);
    switch (location) {
      case Locations.BACKEND.key:
        return this.api().postCredentials(`target.connectUrl == "${targetId}"`, username, password);
      case Locations.BROWSER_SESSION.key:
        this.store.set(targetId, { username, password });
        return of(true);
      default:
        console.warn('Unknown storage location', location);
        return of(false);
    }
  }

  getCredential(targetId: string): Observable<Credential | undefined> {
    const location = getFromLocalStorage('CREDENTIAL_LOCATION', Locations.BACKEND.key);
    switch (location) {
      case Locations.BACKEND.key:
        // if this is stored on the backend then Cryostat should be using those and not prompting us to request from the user
        return of(undefined);
      case Locations.BROWSER_SESSION.key:
        return of(this.store.get(targetId));
      default:
        console.warn('Unknown storage location', location);
        return of(undefined);
    }
  }
}
