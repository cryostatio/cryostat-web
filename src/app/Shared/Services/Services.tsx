/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { NotificationsInstance } from '@app/Notifications/Notifications';
import { TargetService, TargetInstance } from './Target.service';
import { TargetsService } from './Targets.service';
import { ApiService } from './Api.service';
import { NotificationChannel } from './NotificationChannel.service';
import { ReportService } from './Report.service';
import { SettingsService } from './Settings.service';
import { LoginService } from './Login.service';
import { JmxCredentials } from './JmxCredentials.service';

export interface Services {
  target: TargetService;
  targets: TargetsService;
  api: ApiService;
  jmxCredentials: JmxCredentials;
  notificationChannel: NotificationChannel;
  reports: ReportService;
  settings: SettingsService;
  login: LoginService;
}

const settings = new SettingsService();
const jmxCredentials = new JmxCredentials(() => api);
const login = new LoginService(TargetInstance, jmxCredentials, settings);
const api = new ApiService(TargetInstance, NotificationsInstance, login);
const notificationChannel = new NotificationChannel(NotificationsInstance, login);
const reports = new ReportService(login, NotificationsInstance);
const targets = new TargetsService(api, NotificationsInstance, login, notificationChannel);

const defaultServices: Services = {
  target: TargetInstance,
  targets,
  api,
  jmxCredentials,
  notificationChannel,
  reports,
  settings,
  login,
};

const ServiceContext: React.Context<Services> = React.createContext(defaultServices);

export { ServiceContext, defaultServices };
