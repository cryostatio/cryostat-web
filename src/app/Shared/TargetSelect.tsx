/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { LoadingView } from '@app/LoadingView/LoadingView';
import { SerializedTarget } from '@app/Shared/SerializedTarget';
import { ServiceContext } from '@app/Shared/Services/Services';
import { includesTarget, NO_TARGET, Target } from '@app/Shared/Services/Target.service';
import { NoTargetSelected } from '@app/TargetView/NoTargetSelected';
import { getFromLocalStorage } from '@app/utils/LocalStorage';
import { useSubscriptions } from '@app/utils/useSubscriptions';
import {
  Card,
  CardBody,
  CardExpandableContent,
  CardHeader,
  CardTitle,
  Select,
  SelectGroup,
  SelectOption,
  SelectVariant,
} from '@patternfly/react-core';
import { ContainerNodeIcon } from '@patternfly/react-icons';
import * as React from 'react';

export interface TargetSelectProps {
  // display a simple, non-expandable component. set this if the view elsewhere
  // contains a <SerializedTarget /> or other repeated components
  simple?: boolean;
  onSelect?: (target: Target) => void;
}

export const TargetSelect: React.FunctionComponent<TargetSelectProps> = ({ onSelect, simple, ...props }) => {
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();
  const firstLoadRef = React.useRef(false);

  const [isExpanded, setExpanded] = React.useState(false);
  const [selected, setSelected] = React.useState<Target>(NO_TARGET);
  const [targets, setTargets] = React.useState([] as Target[]);
  const [isDropdownOpen, setDropdownOpen] = React.useState(false);
  const [isLoading, setLoading] = React.useState(false);

  const onExpand = React.useCallback(() => {
    setExpanded((v) => !v);
  }, [setExpanded]);

  const _refreshTargetList = React.useCallback(() => {
    setLoading(true);
    addSubscription(
      context.targets.queryForTargets().subscribe(() => {
        // Reset loading and context.targets.targets will emit
        setLoading(false);
      })
    );
  }, [addSubscription, context.targets, setLoading]);

  const handleSelect = React.useCallback(
    (_, selection, isPlaceholder) => {
      setDropdownOpen(false);
      const toSelect: Target = isPlaceholder ? NO_TARGET : selection;
      onSelect && onSelect(toSelect);
      setSelected(toSelect);
    },
    [setDropdownOpen, onSelect, setSelected]
  );

  React.useEffect(() => {
    addSubscription(context.targets.targets().subscribe(setTargets));
  }, [addSubscription, context.targets, setTargets]);

  React.useEffect(() => {
    if (!context.settings.autoRefreshEnabled()) {
      return;
    }
    const id = window.setInterval(
      () => _refreshTargetList(),
      context.settings.autoRefreshPeriod() * context.settings.autoRefreshUnits()
    );
    return () => window.clearInterval(id);
  }, [context.settings, _refreshTargetList]);

  React.useEffect(() => {
    if (selected !== NO_TARGET && !includesTarget(targets, selected)) {
      handleSelect(undefined, undefined, true);
    }
    if (targets.length && !firstLoadRef.current) {
      firstLoadRef.current = true;
      const cachedUrl = getFromLocalStorage('TARGET', undefined);
      const matched = targets.find((tn) => tn.connectUrl === cachedUrl);
      if (matched) {
        handleSelect(undefined, matched, false);
      }
    }
  }, [handleSelect, targets, selected, firstLoadRef]);

  const selectOptions = React.useMemo(() => {
    let options = [] as JSX.Element[];

    const groupNames = new Set<string>();
    targets.forEach((t) => groupNames.add(t.annotations?.cryostat['REALM'] || 'Others'));

    options = options.concat(
      Array.from(groupNames)
        .map((name) => (
          <SelectGroup key={name} label={name}>
            {targets
              .filter((t) => (t.annotations?.cryostat['REALM'] || 'Others') === name)
              .map((t: Target) => (
                <SelectOption key={t.connectUrl} value={t} isPlaceholder={false}>
                  {!t.alias || t.alias === t.connectUrl ? `${t.connectUrl}` : `${t.alias} (${t.connectUrl})`}
                </SelectOption>
              ))}
          </SelectGroup>
        ))
        .sort((a, b) => `${a.props['label']}`.localeCompare(`${b.props['label']}`))
    );
    return options;
  }, [targets]);

  const handleTargetFilter = React.useCallback(
    (_, value: string) => {
      if (!value) {
        return selectOptions;
      }
      const matchExp = new RegExp(value, 'i');
      return selectOptions
        .filter((grp) => grp.props.children)
        .map((grp) =>
          React.cloneElement(grp, {
            children: grp.props.children.filter(
              (option) => matchExp.test(option.props.value.connectUrl) || matchExp.test(option.props.value.alias)
            ),
          })
        )
        .filter((grp) => grp.props.children.length > 0);
    },
    [selectOptions]
  );

  const cardHeaderProps = React.useMemo(
    () =>
      simple
        ? {}
        : {
            onExpand: onExpand,
            toggleButtonProps: {
              id: 'target-select-expand-button',
              'aria-label': 'Details',
              'aria-labelledby': 'expandable-card-title target-select-expand-button',
              'aria-expanded': isExpanded,
            },
          },
    [simple, onExpand, isExpanded]
  );

  return (
    <Card {...props} isRounded isCompact isFlat isExpanded={isExpanded}>
      <CardHeader {...cardHeaderProps}>
        <CardTitle>Target JVM</CardTitle>
      </CardHeader>
      {isLoading ? (
        <LoadingView />
      ) : (
        <>
          <CardBody>
            <Select
              placeholderText="Select a target"
              toggleIcon={<ContainerNodeIcon />}
              variant={SelectVariant.single}
              hasInlineFilter
              inlineFilterPlaceholderText="Filter by target"
              isGrouped
              onFilter={handleTargetFilter}
              onSelect={handleSelect}
              onToggle={setDropdownOpen}
              selections={selected.alias || selected.connectUrl}
              isFlipEnabled={true}
              menuAppendTo="parent"
              maxHeight="20em"
              isOpen={isDropdownOpen}
              aria-label="Select Target"
            >
              {selectOptions}
            </Select>
          </CardBody>
          <CardExpandableContent>
            <CardBody>
              {selected === NO_TARGET ? <NoTargetSelected /> : <SerializedTarget target={selected} />}
            </CardBody>
          </CardExpandableContent>
        </>
      )}
    </Card>
  );
};
