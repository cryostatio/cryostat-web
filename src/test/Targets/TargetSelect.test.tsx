/*
 * Copyright The Cryostat Authors
 * 
 * The Universal Permissive License (UPL), Version 1.0
 * 
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 * 
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 * 
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 * 
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import renderer from 'react-test-renderer'
import { cleanup, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { of } from 'rxjs';

import { CUSTOM_TARGETS_REALM, TargetSelect } from '@app/TargetSelect/TargetSelect';
import { defaultServices, ServiceContext } from '@app/Shared/Services/Services';
import userEvent from '@testing-library/user-event';
import { Target } from '@app/Shared/Services/Target.service';

const mockFooConnectUrl = 'service:jmx:rmi://someFooUrl';
const mockBarConnectUrl = 'service:jmx:rmi://someBarUrl';
const mockBazConnectUrl = 'service:jmx:rmi://someBazUrl';

// Test fails if new Map([['REALM', 'Custom Targets']]) is used, most likely since 'cryostat' Map is not being utilized
const cryostatAnnotation = new Map();
cryostatAnnotation['REALM'] = CUSTOM_TARGETS_REALM;
const mockFooTarget: Target = { 
    connectUrl: mockFooConnectUrl, 
    alias: 'fooTarget', 
    annotations: { 
        cryostat: cryostatAnnotation, 
        platform : new Map() 
    }
};
const mockBarTarget: Target = { ...mockFooTarget, connectUrl: mockBarConnectUrl, alias: 'barTarget' }
const mockBazTarget: Target = { connectUrl: mockBazConnectUrl, alias: 'bazTarget' }

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useRouteMatch: () => ({ url: '/baseUrl' }),
  }));

jest.spyOn(defaultServices.target, 'target')
    .mockReturnValue(of(mockFooTarget))
    .mockReturnValueOnce(of())
    .mockReturnValueOnce(of())
    .mockReturnValueOnce(of(mockFooTarget))
    .mockReturnValueOnce(of(mockFooTarget))
    .mockReturnValueOnce(of(mockFooTarget))
    .mockReturnValueOnce(of(mockBazTarget)) // does nothing when trying to delete non-custom targets

jest.spyOn(defaultServices.targets, 'targets')
    .mockReturnValue(of([mockFooTarget, mockBarTarget]))
    .mockReturnValueOnce(of([mockFooTarget]))
    .mockReturnValueOnce(of([mockFooTarget]))
    .mockReturnValueOnce(of([mockFooTarget, mockBarTarget])) // renders dropdown of multiple discovered targets
;

jest.spyOn(defaultServices.api, 'createTarget')
    .mockReturnValue(of(true));

jest.spyOn(defaultServices.api, 'deleteTarget')
    .mockReturnValue(of(true));

jest.spyOn(defaultServices.targets, 'queryForTargets')
    .mockReturnValue(of());

jest.spyOn(defaultServices.notificationChannel, 'messages')
    .mockReturnValue(of())


afterEach(cleanup);

describe('<TargetSelect />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(    
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('contains the correct information', () => {
    render(
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>
    );

    expect(screen.getByText('Target JVM')).toBeInTheDocument();
    expect(screen.getByText(`Select Target...`)).toBeInTheDocument();

    expect(screen.getByLabelText('Create target')).toBeInTheDocument();
    expect(screen.getByLabelText('Delete target')).toBeInTheDocument();
    expect(screen.getByLabelText('Refresh targets')).toBeInTheDocument();
    expect(screen.getByLabelText('Options menu')).toBeInTheDocument();
  });

  it('renders dropdown of multiple discovered targets', () => {
    render(
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>
    );

    expect(screen.getByText(`fooTarget`)).toBeInTheDocument();
    
    userEvent.click(screen.getByLabelText("Options menu"));
    expect(screen.getByLabelText('Select Input')).toBeInTheDocument();
    expect(screen.getByText(`Select Target...`)).toBeInTheDocument();
    expect(screen.getByText(`fooTarget (service:jmx:rmi://someFooUrl)`)).toBeInTheDocument();
    expect(screen.getByText(`barTarget (service:jmx:rmi://someBarUrl)`)).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument(); // Number of discoverable targets
  });

  it('creates a target if user completes modal', () => {
    render(
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>
    );
    const createButton = screen.getByLabelText('Create target');
    userEvent.click(createButton);

    const textBoxes = screen.getAllByRole('textbox');

    userEvent.type(textBoxes[0], 'service:jmx:rmi://someBazUrl');
    userEvent.type(textBoxes[1], 'bazTarget');

    const createTargetRequestSpy = jest.spyOn(defaultServices.api, 'createTarget');
    userEvent.click(screen.getByText('Create'));

    expect(createTargetRequestSpy).toBeCalledTimes(1);
    expect(createTargetRequestSpy).toBeCalledWith(mockBazTarget);
  }); 

  it('deletes target when delete button clicked', async () => {
    render(
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>
    );

    const deleteButton = screen.getByLabelText('Delete target');
    await waitFor(() => expect(deleteButton).not.toBeDisabled());

    userEvent.click(deleteButton);

    const deleteTargetRequestSpy = jest.spyOn(defaultServices.api, 'deleteTarget');
    expect(deleteTargetRequestSpy).toBeCalledTimes(1);
    expect(deleteTargetRequestSpy).toBeCalledWith(mockFooTarget);
  }); 

  it('does nothing when trying to delete non-custom targets', () => {
    render(
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>
    );
    const deleteButton = screen.getByLabelText('Delete target');
    userEvent.click(deleteButton);

    const deleteTargetRequestSpy = jest.spyOn(defaultServices.api, 'deleteTarget');

    expect(deleteTargetRequestSpy).toBeCalledTimes(0);
    expect(deleteButton).toBeDisabled();
  }); 

  it('refreshes targets when button clicked', () => {
    render(
        <ServiceContext.Provider value={defaultServices}>
            <TargetSelect />
        </ServiceContext.Provider>
    );
    const refreshButton = screen.getByLabelText('Refresh targets');
    userEvent.click(refreshButton);

    const refreshTargetsRequestSpy = jest.spyOn(defaultServices.targets, 'queryForTargets');
    expect(refreshTargetsRequestSpy).toBeCalledTimes(1);
  }); 
});
