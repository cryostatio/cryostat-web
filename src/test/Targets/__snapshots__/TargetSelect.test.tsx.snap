// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TargetSelect /> renders correctly 1`] = `
<article
  className="pf-c-card"
  data-ouia-component-id="OUIA-Generated-Card-1"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe={true}
  id=""
>
  <div
    className="pf-c-card__header"
  >
    <div
      className=""
    >
      <h4
        className=""
        data-ouia-component-id="OUIA-Generated-Text-1"
        data-ouia-component-type="PF4/Text"
        data-ouia-safe={true}
        data-pf-content={true}
      >
        Target JVM
      </h4>
    </div>
    <div
      className="pf-c-card__actions"
    >
      <button
        aria-disabled={false}
        aria-label="Create target"
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        <span
          className="pf-c-button__icon pf-m-start"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            />
          </svg>
        </span>
      </button>
      <button
        aria-disabled={true}
        aria-label="Delete target"
        className="pf-c-button pf-m-control pf-m-disabled"
        data-ouia-component-id="OUIA-Generated-Button-control-2"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={true}
        onClick={[Function]}
        role={null}
        tabIndex={null}
        type="button"
      >
        <span
          className="pf-c-button__icon pf-m-start"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </span>
      </button>
      <button
        aria-disabled={false}
        aria-label="Refresh targets"
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-3"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        <span
          className="pf-c-button__icon pf-m-start"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M515.8,0 C655.3,0 781.6,58.4 873.4,150.2 L968.9,54.7 C989,34.5 1024,48.8 1024,77.3 L1024,352 C1024,369.7 1009.2,384.1 991.5,384 L716.8,384 C688.3,384 674,349.5 694.2,329.3 L783,240.5 C713.5,171.1 617.8,128 512,128 C300.4,128 128,300.2 128,512 C128,723.8 300.2,895.1 511.9,895 C606.9,895 693.9,861.2 761,803.9 C773.7,793.1 792.6,793.8 804.4,805.6 L849.6,850.9 C862.6,864 861.9,885.4 848,897.6 C758,976.3 640.3,1024 511.7,1024 C229.1,1024 0,797 0,510.7 C0,226.2 235,0 515.8,0"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="pf-c-card__body"
  >
    <div
      className="pf-c-select"
      data-ouia-component-id="OUIA-Generated-Select-single-1"
      data-ouia-component-type="PF4/Select"
      data-ouia-safe={true}
    >
      <button
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-label="Options menu"
        aria-labelledby=" pf-select-toggle-id-1"
        className="pf-c-select__toggle"
        disabled={false}
        id="pf-select-toggle-id-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <div
          className="pf-c-select__toggle-wrapper"
        >
          <span
            className="pf-c-select__toggle-icon"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M22.3,576.00143 L1001.8,576.00143 C1013.2,576.00143 1020.3,570.90143 1023.1,560.50143 C1025.6,550.20143 1022.9,537.30143 1014.4,522.10143 C1014.4,522.10143 874.3,232.50143 863.8,211.00143 C856,191.90143 841.4,192.000713 841.4,192.000713 L185.6,192.000713 C185.6,192.000713 169,191.60143 161.8,209.70143 C151.3,232.30143 9.7,522.00143 9.7,522.00143 C1.2,537.10143 -1.7,550.00143 1,560.40143 C3.7,570.80143 10.8,576.00143 22.3,576.00143 Z M1013.1,650.90143 C1005.8,643.60143 997.2,640.00143 987.4,640.00143 L36.6,640.00143 C26.6,640.00143 18.2,643.60143 10.9,650.90143 C3.6,658.20143 0,666.80143 0,676.60143 L0,795.50143 C0,805.50143 3.6,813.90143 10.7,821.20143 C18,828.40143 26.4,832.00143 36.4,832.00143 L987.4,832.00143 C997.4,832.00143 1005.8,828.40143 1013.1,821.10143 C1020.4,813.80143 1024,805.40143 1024,795.40143 L1024,676.60143 C1024,666.60143 1020.4,658.20143 1013.1,650.90143 Z M128,768.00143 L64,768.00143 L64,704.00143 L128,704.00143 L128,768.00143 Z M256,768.00143 L192,768.00143 L192,704.00143 L256,704.00143 L256,768.00143 Z M384,768.00143 L320,768.00143 L320,704.00143 L384,704.00143 L384,768.00143 Z M512,768.00143 L448,768.00143 L448,704.00143 L512,704.00143 L512,768.00143 Z M960,768.00143 L896,768.00143 L896,704.00143 L960,704.00143 L960,768.00143 Z"
              />
            </svg>
          </span>
          <span
            className="pf-c-select__toggle-text"
          >
            Select Target...
          </span>
        </div>
        <span
          className="pf-c-select__toggle-arrow"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</article>
`;
